src/gui/qgsnewvectortabledialog.cpp

```diff
@@ -0,0 +1,753 @@
+  mGeomTypeCbo->addItem( QgsApplication::getThemeIcon( QStringLiteral( "mIconTableLayer.svg" ) ), tr( "No Geometry" ), QString() );
+  mGeomTypeCbo->addItem( QgsApplication::getThemeIcon( QStringLiteral( "mIconPointLayer.svg" ) ), tr( "Point" ), QgsWkbTypes::Type::Point );
+  mGeomTypeCbo->addItem( QgsApplication::getThemeIcon( QStringLiteral( "mIconLineLayer.svg" ) ), tr( "Line" ), QgsWkbTypes::Type::LineString );
+  mGeomTypeCbo->addItem( QgsApplication::getThemeIcon( QStringLiteral( "mIconPolygonLayer.svg" ) ), tr( "Polygon" ), QgsWkbTypes::Type::Polygon );
+  mGeomTypeCbo->addItem( QgsApplication::getThemeIcon( QStringLiteral( "mIconPointLayer.svg" ) ), tr( "MultiPoint" ), QgsWkbTypes::Type::MultiPoint );
+  mGeomTypeCbo->addItem( QgsApplication::getThemeIcon( QStringLiteral( "mIconLineLayer.svg" ) ), tr( "MultiLine" ), QgsWkbTypes::Type::MultiLineString );
+  mGeomTypeCbo->addItem( QgsApplication::getThemeIcon( QStringLiteral( "mIconPolygonLayer.svg" ) ), tr( "MultiPolygon" ), QgsWkbTypes::Type::MultiPolygon );
+  mGeomTypeCbo->setCurrentIndex( 0 );
+
+  connect( mFieldsTableView->selectionModel(), &QItemSelectionModel::selectionChanged, mFieldsTableView, [ = ]( const QItemSelection & selected, const QItemSelection & )
+  {
+    if ( ! selected.isEmpty() )
+    {
+      mCurrentRow = selected.indexes().first().row();
+    }
+    updateButtons();
+  } );
+
+  // Get a default type for new fields
+  const QVariant::Type defaultFieldType { mFieldModel->nativeTypes().first().mType };
+  const QString defaultFieldTypeName { mFieldModel->nativeTypes().first().mTypeName };
+
+  // Actions
+  connect( mAddFieldBtn, &QPushButton::clicked, this, [ = ]
+  {
+    QgsFields fieldList { fields() };
+    QgsField newField { QStringLiteral( "new_field_name" ), defaultFieldType, defaultFieldTypeName };
+    fieldList.append( newField );
+    setFields( fieldList );
+    selectRow( fieldList.count() - 1 );
+  } );
+
+  connect( mDeleteFieldBtn, &QPushButton::clicked, this, [ = ]
+  {
+    QgsFields fieldList { fields() };
+    if ( fieldList.exists( mCurrentRow ) )
+    {
+      fieldList.remove( mCurrentRow );
+      setFields( fieldList );
+      mCurrentRow = -1;
+    }
+  } );
+
+  connect( mFieldUpBtn, &QPushButton::clicked, this, [ = ]
+  {
+    if ( fields().exists( mCurrentRow ) && fields().exists( mCurrentRow - 1 ) )
+    {
+      QgsFields fieldList;
+      for ( int i = 0; i < fields().count(); ++i )
+      {
+        if ( i == mCurrentRow - 1 )
+        {
+          fieldList.append( fields().at( mCurrentRow ) );
+          fieldList.append( fields().at( mCurrentRow - 1 ) );
+        }
+        else if ( i != mCurrentRow )
+        {
+          fieldList.append( fields().at( i ) );
+        }
+      }
+      setFields( fieldList );
+      selectRow( mCurrentRow - 1 );
+    }
+  } );
+
+  connect( mFieldDownBtn, &QPushButton::clicked, this, [ = ]
```

Comment:
I like this approach of putting the lambdas directly in connections for button actions like this, good move

Reply:
Yeah, I was asking myself if I was abusing it but I like to have the action code close to the connection, I guess the only reason to switch the lambda with an independent function is if it's called from multiple places.